lib/core/networking/websockets/session.dart

```diff
@@ -0,0 +1,178 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'middleware.dart';
+
+/// Base class for client and server sessions
+abstract class Session<T> {
+  /// Awaitable to ensure ready for use
+  late final Future<void> ready;
+
+  /// Individial storage attached to each session for ephemeral data
+  Map<dynamic, dynamic> storage = <dynamic, dynamic>{};
+
+  /// WebSocket connection used to transfer data from client <-> server
+  late final WebSocket _socket;
+
+  /// A reference to the instantiated subclass.
+  /// This is used to allow passing of non base [Session] objects
+  /// such as [ServerSession] or [ClientSession].
+  T? _ref;
+
+  /// Handler for received messages
+  Future<dynamic> Function(T, dynamic)? onMessage;
+
+  /// Handler for encountered errors
+  void Function(T, Object)? onError;
+
+  /// Handler for stream closure
+  void Function(T)? onDone;
+
+  /// Send data over the WebSocket
+  Future<void> send(dynamic message) async {
+    // Wait for the connection to be established to avoid errors
+    await ready;
+
+    if (_socket.readyState == WebSocket.open) _socket.add(message);
+  }
+
+  /// Teardown the connection and remove references to enable garbage collection
+  Future<void> close() async {
+    // Wait for the connection to be established to avoid errors
+    await ready;
+
+    await _socket.close();
+  }
+
+  /// Proxy to call the internal [_onError] method.
+  ///
+  /// This will cause the connection to be closed.
+  /// [error] will be passed to the [onError] handler.
+  void raise(Object error) => _onError(error);
+
+  /// Listen on the WebSocket.
+  /// This should be called AFTER the WebSocket has been created by the subclass.
+  void _listen() => _socket.listen(_onData, onError: _onError, onDone: _onDone);
+
+  /// Internal delegation handler for received messages
+  void _onData(dynamic message) async {
+    if (onMessage != null) {
+      final dynamic response = await onMessage!(_ref!, message);
+      if (response != null) send(response);
+    }
+  }
+
+  /// Internal delegation handler for encountered errors
+  void _onError(Object error) async {
+    if (onError != null) {
+      onError!(_ref!, error);
+    }
+
+    await close();
+  }
+
+  /// Internal delegation handler for stream closure
+  void _onDone() {
+    if (onDone != null) {
+      onDone!(_ref!);
+    }
+
+    storage.clear();
+    _ref = null;
+    onMessage = null;
+    onError = null;
+  }
+}
+
+/// Extends base session to add support for installing middleware
+class MiddlewareSession<T> extends Session<T> {
+  /// List of installed middleware.
+  ///
+  /// Each middleware recieves the calling [Session] instance, message,
```

Comment:
```suggestion
  /// Each middleware receives the calling [Session] instance, message,
```

Please use vs code to fix all of those :)